### ISSUE TITLE
Impossible to use combine(Iterable<Flow>) with type params from non-inlined call site

### ISSUE THREAD

**Issue by BenTilbrook** [Kotlin/kotlinx.coroutines#2945]
It's not possible to use `combine<Iterable<Flow<T>>` with type parameters unless the call site is also inlined, supplying the `T` type param as reified.

Example:

```kotlin
// Does not compile
fun <N : Number> first(flows: Iterable<Flow<N>>) = combine(flows) { array -> array.first() }
```

This example is trivial for brevity, but in many cases it's not at all practical to make the calling function `inline`.

```
Cannot use 'N' as reified type parameter. Use a class instead.
```

This is a significant and unnecessary limitation. If `List` was used instead of `Array`, or an alternative `combine` function using `List` was available, this wouldn't be a problem.

It's not clear why `Array` is in use at all. I assume it's for performance.

Since `combineInternal` is `internal`, we have no choice but to copy its entire implementation in order to use it in the above scenario.

**Comment by qwwdfsad** [comment]
Thanks for the report!

The array was used for the sake of performance and to "trick" the JVM type system. It probably can be revisited, though it may be too harsh for compatibility.

>Since combineInternal is internal, we have no choice but to copy its entire implementation in order to use it in the above scenario.

There exists simpler workaround
```
fun <N : Number> first(flows: Iterable<Flow<N>>) = combine(flows as Iterable<Flow<Any>>) { array -> array.first() as N }
```

**Comment by BenTilbrook** [comment]
Thank you. Good workaround to cast to `Any` - should've thought of that!

**Comment by qwwdfsad** [comment]
We've discussed it internally and, unfortunately, we cannot fix it. Such change (`Array` -> `List`) is impossible to make non-breaking, and taking into account the number of overloads (`combine`, `combineTransform`, and most of them without an array in the signature), the new name is also not an option.
Alas, we decided to keep it as is and consider it our legacy.

But there is a workaround without unchecked casts, specifying the type explicitly:
`fun <N : Number> first(flows: Iterable<Flow<N>>) = combine<Number, Number>(flows) { array -> array.first() }`
And we also expect IDE to assist here as well: https://youtrack.jetbrains.com/issue/KTIJ-19759, so in the future it would be easier for users to adapt
